### PR TITLE
Run "pkg sync" before running "inko build" if needed

### PR DESCRIPTION
**Pull request by uasi**
Closes https://github.com/inko-lang/inko/issues/722

Followed all steps described in #722.

### CODE REVIEW

File: inko/src/pkg/sync.rs [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        )
+    })?;
+
+    Ok(blake3::hash(&data).to_hex().to_string())
+}
+
+fn save_manifest_hash(directory: &Path) -> Result<(), String> {
+    if !Path::new(MANIFEST_FILE).exists() {
+        return Ok(());
+    }
```
Comment (yorickpeterse) [inline inko/src/pkg/sync.rs:201]: I think this check isn't needed. This method is only called in `sync()`, which in turn is only called if either `inko.pkg` changed (meaning it must exist in the first place), or if one runs `inko pkg sync`.

I think what we can do is the following:

We change `manifest_hash` such that if `read()` returns an error, we check if that error is `std::io::ErrorKind::NotFound`. If so, we return `Ok(None)`. For any other error we return `Err(String)`. If the file does exist, we return `Ok(Some(the hash))`.

In `save_manifest_hash` we then call `manifest_hash`, and if it returns `Ok(None)` we just return early. This way we don't need the extra `exists()` check.

**Review by yorickpeterse — CHANGES_REQUESTED** [review]
Overall this looks great, just a few small comments :smiley:

**Comment by yorickpeterse** [comment]
Oh I forgot to add: we should also call `sync_if_needed` when running the following commands:

- `inko check`
- `inko doc`
- `inko test`

File: inko/src/pkg/sync.rs [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+    Ok(())
+}
+
+fn manifest_hash_changed(directory: &Path) -> Result<bool, String> {
+    let hash = match manifest_hash() {
+        Ok(Some(hash)) => hash,
+        // If the manifest file doesn't exist, we assume it has changed if the
+        // dependencies directory exists, as it indicates that the user once
+        // had a manifest file but deleted it when it was no longer needed.
```
Comment (uasi) [inline inko/src/pkg/sync.rs:239]: > the user once had a manifest file but deleted it when it was no longer needed

In this case, empty inko.pkg will be re-created on running pkg sync.

**Review by yorickpeterse — APPROVED** [review]

**Comment by yorickpeterse** [comment]
@uasi Thanks! :tada:

**Pull request merged**